Update configs on block finalization

dabasov opened the pull request:
Update Config on block finalization. Config will be updated in case there is new version of config in block's state

dabasov committed:
Update Config on block finalization. Config will be updated in case there is new version of config in block's state

peterlimg commented:
Seems the `config` is not thread safe, if we update it in the `SetLatestFinalizedBlock()`, we'd better protect the config with locks as the `SetLatestFinalizedBlock()` could be called in different places concurrently.

Sriep commented on code/go/0chain.net/chaincore/chain/entity.go:
```diff
@@ -1367,6 +1370,68 @@ func (c *Chain) SetLatestFinalizedBlock(b *block.Block) {
+	clientState, err := mc.getClientState(pb)
+	if err != nil {
+		// This might happen after stopping and starting the miners
+		// and the MPT has not been setup yet.
```
Does this handle https://github.com/0chain/0chain/issues/541, if so can we link the issue to this Pr?

dabasov committed:
synchronized access to Config

dabasov committed:
Merge branch 'staging' into fix/update-configs-on-finalization

# Conflicts:
#	code/go/0chain.net/miner/protocol_block_main.go

dabasov commented:
@peterlimg Added synchronized access to config structure:
1. Extracted interface
2. Moved json structure to DTO, because it is not possible to parse json to not exported fields. 
3. Made Config structs not exportable, to guard access with RW lock
4. All updates to DTO is made through write lock

It is important to notice, that working to struct is still not fully concurrent safe, if we need to change or read several properties atomically we must guard such action with lock. 
Probably, it is not needed  for writing now and most likely won't be needed in the future until update will be made outside MPT, as now we make all changes through blockchain transaction which is atomic by nature. 
Reads can still need  some atomicity, when several properties depend on each other. Such atomicity can be achieved by extracting such dependant properties to single structure.

@peterlimg @Sriep do you know such dependant properties?

dabasov committed:
unit test fix

peterlimg committed:
Merge branch 'staging' into fix/update-configs-on-finalization

peterlimg committed:
Merge branch 'staging' into fix/update-configs-on-finalization

dabasov committed:
Merge branch 'staging' into fix/update-configs-on-finalization

dabasov committed:
Merge branch 'staging' into fix/update-configs-on-finalization

dabasov committed:
Merge branch 'staging' into fix/update-configs-on-finalization

dabasov committed:
Merge branch 'staging' into fix/update-configs-on-finalization
